libraries/adb/src/server/client.ts

```diff
@@ -0,0 +1,391 @@
+            const status = parts[1]!;
+            if (status !== "device") {
+                continue;
+            }
+
+            let product: string | undefined;
+            let model: string | undefined;
+            let device: string | undefined;
+            let transportId: number | undefined;
+            for (let i = 2; i < parts.length; i += 1) {
+                const [key, value] = parts[i]!.split(":");
+                switch (key) {
+                    case "product":
+                        product = value;
+                        break;
+                    case "model":
+                        model = value;
+                        break;
+                    case "device":
+                        device = value;
+                        break;
+                    case "transport-id":
+                        transportId = Number.parseInt(value!, 10);
+                        break;
+                }
+            }
+            if (!transportId) {
+                throw new Error(`No transport id for device ${serial}`);
+            }
+            const features = await this.getDeviceFeatures({ transportId });
+            const banner = new AdbBanner(product, model, device, features);
+            devices.push(
+                new AdbServerTransport(this, serial, banner, transportId)
+            );
+        }
+        connection.writable.close().catch(NOOP);
+        readable.cancel().catch(NOOP);
+        return devices;
+    }
+
+    public formatDeviceService(
+        device: AdbServerDeviceSelector,
+        command: string
+    ) {
+        if (!device) {
+            return `host:${command}`;
+        }
+        if ("transportId" in device) {
+            return `host-transport-id:${device.transportId}:${command}`;
+        }
+        if ("serial" in device) {
+            return `host-serial:${device.serial}:${command}`;
+        }
+        if ("usb" in device) {
+            return `host-usb:${command}`;
+        }
+        if ("emulator" in device) {
+            return `host-local:${command}`;
+        }
+        throw new Error("Invalid device selector");
+    }
+
+    public async getDeviceFeatures(
+        device: AdbServerDeviceSelector
+    ): Promise<AdbFeature[]> {
+        const connection = await this.connect(
+            this.formatDeviceService(device, "features")
+        );
+        const readable = new BufferedReadableStream(connection.readable);
+        try {
+            const response = await AdbServerClient.readString(readable);
+            return response.split(",") as AdbFeature[];
+        } finally {
+            connection.writable.close().catch(NOOP);
+            readable.cancel().catch(NOOP);
+        }
+    }
+
+    public async createDeviceSocket(
+        device: AdbServerDeviceSelector,
+        service: string
+    ): Promise<AdbSocket> {
+        let switchService: string;
+        if (!device) {
+            switchService = `host:transport-any`;
```

Comment:
Switch to new `tport:` service? (What's the benefit of doing that?)

Reply:
We can use `host:tport:any` then `host:features` to acquire transport id and device features.

In theory we should not send host commands after switching to device socket, but adb server does support this.

It's always possible to send these commands in two connections (adb server switch connection to device socket after receving device command, but if we don't send the device command, it should be zero cost)